### PR TITLE
Add Birthcry event to kube-proxy

### PR DESCRIPTION
**Pull request by jiangyaoguo**
#7304
1. Add HostnameOverride parameter for kube-proxy as kubelet did.
2. Add Birthcry event for kube-proxy.

### CODE REVIEW

**Comment by gmarek** [comment]
cc @wojtek-t 
Currently events are one of the things that put most pressure on the master. We should carefully consider which ones to add.

**Comment by jiangyaoguo** [comment]
@gmarek @erictune Because we should carefully consider adding events, I  plan to add basic birthcry event firstly and then make other events pr for discuss.

**Comment by gmarek** [comment]
cc @timothysc and assigning @erictune as he opened the original issue.

**Comment by timothysc** [comment]
I would hope this would have minimal impact, I'm not concerned, unless we start expanding the noise.

**Comment by gmarek** [comment]
@timothysc I just believe that we should have someone who'll monitor all Master<->Node communication changes. Like all API changes need to go through Brian. I'm volunteering you and @wojtek-t :)

**Comment by timothysc** [comment]
@gmarek we should create the working-group labels, then we'll always have visibility ;-)

**Comment by thockin** [comment]
There was other work afoot to rate-limit the impact of this, but yeah - it
makes me nervous that a network event could cause a synchronized thundering
herd.

File: cmd/kube-proxy/app/server.go [hunk 1]
```diff
@@ -50,6 +54,10 @@ type ProxyServer struct {
 	Master             string
 	Kubeconfig         string
 	PortRange          util.PortRange
+	Recoder            record.EventRecorder
```
Comment (bgrant0607) [inline cmd/kube-proxy/app/server.go]: Recorder

**Comment by bgrant0607** [comment]
Thundering herd: We have that problem already, when all kube-proxies try to reconnect to apiserver at the same time.

Mitigating impact of crashlooping kube-proxies: Shouldn't event compression apply here also?

Specific event: Should be consistent with other events of the same purpose. Event.Reason should be CamelCase (starting with a capital letter), as with Reason everywhere in the API.

cc @saad-ali

File: cmd/kube-proxy/app/server.go [hunk 2]
```diff
@@ -151,3 +176,7 @@ func (s *ProxyServer) Run(_ []string) error {
 	proxier.SyncLoop()
 	return nil
 }
+
+func (s *ProxyServer) BirthCry() {
```
Comment (timothysc) [inline cmd/kube-proxy/app/server.go]: Will this ever be called in any other control path then the one on line 135?  if not then collapse.

**Comment by saad-ali** [comment]
> Mitigating impact of crashlooping kube-proxies: Shouldn't event compression apply here also

Because the event compression is done client side, if the binary is terminated, the event cache is gone, and compression will not happen.

**Comment by wojtek-t** [comment]
I think that this discussion leads us again to the point that we should really seriously consider moving events to a separate storage (either another instance of etcd, or something else).

**Comment by jiangyaoguo** [comment]
Maybe we need a event aggregator in the future. @wojtek-t Does saving events to another etcd through apiserver will be helpful to reduce pressure of apiserver?

**Comment by timothysc** [comment]
@jiangyaoguo yes.  

I debate the cost vs. benefit of routing events through the api-server... at all. 
Events are not for consumption, and should not be used for actionable cluster state...

**Comment by wojtek-t** [comment]
>
I debate the cost vs. benefit of routing events through the api-server... at all. 
Events are not for consumption, and should not be used for actionable cluster state...

I agree with the first sentence.
However, in my opinion events are for consumption. There can be useful e.g. for debugging. However, this doesn't mean that we should store them in etcd.

In my opinion events are "kind of logs" and we should treat the similarly. We do not store logs in etcd, and we probably shouldn't store there events too.

**Comment by gmarek** [comment]
IIUC Tims point was that is wasn't for the consumption for any of the clusters system. They are for human consumption only.

**Comment by timothysc** [comment]
@gmarek yup

**Comment by erictune** [comment]
This is just making the kube-proxy behave the same as the kubelet.

**Comment by erictune** [comment]
ok to test

**Comment by erictune** [comment]
lgtm

**Comment by wojtek-t** [comment]
@gmarek @timothysc - thanks, I didn't understand your point.

So it basically seems that we're on the same page with that.

**Comment by erictune** [comment]
I agree with Tim's suggestion that we need different storage for events.  But I don't think we block all new events on that effort.

**Comment by erictune** [comment]
#4432 is an issue about moving events out of etcd.

**Pull request merged**